### ISSUE TITLE
Collector Panic

### ISSUE THREAD

**Issue by mamoep** [NetApp/harvest#174]
**Describe the bug**
Collector panic occurs when MetroCluster FC backend fabric goes down.

**Environment**
- Harvest version: harvest version 21.05.2-1 (commit ce091de) (build date 2021-06-14T20:31:09+0530) linux/amd64
- Command line arguments used: [e.g. `bin/harvest start --config=foo.yml --collectors Zapi`]
- OS: RHEL 7.9
- Install method: yum
- ONTAP Version: 9.7P13
- Other:

**To Reproduce**
- Monitor MetroCluster FC with Harvest
- do "switchdisable" on one Brocade Backend FC Switch

**Expected behavior**
Collector should keep running

**Actual behavior**
Collector panics

```{
    "level": "error",
    "Poller": "REDACTED",
    "collector": "ZapiPerf:Path",
    "stack": [
        {
            "func": "New",
            "line": "35",
            "source": "errors.go"
        },
        {
            "func": "(*AbstractCollector).Start.func1",
            "line": "241",
            "source": "collector.go"
        },
        {
            "func": "gopanic",
            "line": "965",
            "source": "panic.go"
        },
        {
            "func": "goPanicIndex",
            "line": "88",
            "source": "panic.go"
        },
        {
            "func": "(*MetricFloat64).SetValueFloat64",
            "line": "97",
            "source": "metric_float64.go"
        },
        {
            "func": "(*ZapiPerf).PollData",
            "line": "319",
            "source": "zapiperf.go"
        },
        {
            "func": "(*task).Run",
            "line": "60",
            "source": "schedule.go"
        },
        {
            "func": "(*AbstractCollector).Start",
            "line": "270",
            "source": "collector.go"
        },
        {
            "func": "goexit",
            "line": "1371",
            "source": "asm_amd64.s"
        }
    ],
    "error": "goroutine panic => goroutine 392 [running]:
runtime/debug.Stack(0xc000419440, 0xab9203, 0x0)
\truntime/debug/stack.go:24 +0x9f
goharvest2/cmd/poller/collector.(*AbstractCollector).Start.func1(0xc000274790)
\tgoharvest2/cmd/poller/collector/collector.go:241 +0x9d
panic(0xab92c0, 0xc000231050)
\truntime/panic.go:965 +0x1b9
goharvest2/pkg/matrix.(*MetricFloat64).SetValueFloat64(0xc000f5c320, 0xc0017e9890, 0x41d83213587fa2a7, 0xc000bfb980, 0x1)
\tgoharvest2/pkg/matrix/metric_float64.go:97 +0x6c
goharvest2/cmd/collectors/zapiperf.(*ZapiPerf).PollData(0xc000173b20, 0x0, 0x0, 0x0)
\tgoharvest2/cmd/collectors/zapiperf/zapiperf.go:319 +0x1baa
goharvest2/cmd/poller/schedule.(*task).Run(0xc0004ccc40, 0x2306eebe8ec, 0xe28da0, 0xdf879b9f8)
\tgoharvest2/cmd/poller/schedule/schedule.go:60 +0x5a
goharvest2/cmd/poller/collector.(*AbstractCollector).Start(0xc000274790, 0xc000cb05b0)
\tgoharvest2/cmd/poller/collector/collector.go:270 +0x1b5
created by main.(*Poller).Start
\tcommand-line-arguments/poller.go:313 +0x225
",
    "caller": "goharvest2/cmd/poller/collector/collector.go:241",
    "time": "2021-06-15T08:49:05+02:00",
    "message": "Collector panicked"
}
```

**Possible solution, workaround, fix**
Panic recovery works, process keeps running.

**Comment by cgrinds** [comment]
I suspect this is similar to https://github.com/NetApp/harvest/issues/160
at line 319, we trying to record the timestamp value of instance

```
318  // add batch timestamp as custom counter
319  if err := timestamp.SetValueFloat64(instance, ts); err != nil {
320  	me.Logger.Error().Stack().Err(err).Msg("set timestamp value: ")
321  }
```

the panic happens at line 97 (below) because i.index runs off the end of me.record
in other words i.index >= len(me.record)

```
96  func (me *MetricFloat64) SetValueFloat64(i *Instance, v float64) error {
97  	me.record[i.index] = true
```

this happened in #160 because the instance indexes changed due to rereading harvest.yml. In this case, the instance indexes are probably wrong due the disabled switch causing instance removals without a corresponding index change.

These metric `setvalue*` methods should do bounds checking and not panic, but the root cause is likely in the caller, zapiperf, using invalid instance indexes.

**Comment by hashi825** [comment]
Unrelated to the exact way to reproduce this issue but I've noticed with collector panics, is that they say they recover but stop actually collecting data. I've noticed panics with the `ZapiPerf:Volume` collector, you'll have log entries after the fact that its behind schedule and appears to be working but pretty much no volume metrics from that relevant cluster after that point, and that poller has to be restarted to work again. Other collectors on that poller continue to work when this occurs.
